### PR TITLE
CBG-3861 support updating vv on xdcr

### PR DESCRIPTION
**Pull request by torcolvin**
- this code will only update cv when appropriate, does not yet support changing pv or mv (not supported in mobile xdcr anyway)
- stubs out some kinds of tests we can write
- Creates a `SourceID` for rosmar that is just bucketUUID (clusterUUID is always empty string), but this will be unique since rosmar is only in memory
- replace `cblEncodeTestSources` with `EncodeTestHistory` since these were duplicate functions and I was about to write a third.

Parts of this PR I'm particularly interested in are whether the test code is easily readable. I gave some limited thought to how to parameterize `TestVVWriteOnDestAfterInitialReplication` and `TestVVWriteTwice` like `TestReplicateVV`, but I couldn't think of something that wasn't like:

- test.preXDCRFunc1
- run xdcr
- test.assertVV1
- test.preXDCRFunc2
- test.assertVV2

I didn't feel like this was more readable and generalizable enough. I am imagining that we have a different set of test cases for the future when we implement `pv` support that doesn't take a datastore, but rather has two `DatabaseContext` instances which would allow us to use the crud operations on `DatabaseContext` and also creating imports. When implementing this, I was going to debate if a structure like above was appropriate.

Ultimately, I just decided to stop working on this to be able to get some code in place and be able to break tasks up into smaller pieces. I do think it's worthwhile to get this in good shape and any significant improvements, I'll break into separate tickets.

### CODE REVIEW

File: xdcr/xdcr_test.go [hunk 1]
```diff
@@ -50,74 +52,299 @@ func TestMobileXDCRNoSyncDataCopied(t *testing.T) {
...
+			preXDCRFunc: func(t *testing.T, docID string) uint64 {
+				_, err := toDs.WriteCas(docID, 0, 0, []byte(`{"datastore":"toDs"}`), 0)
+				require.NoError(t, err)
+				cas, err := fromDs.WriteCas(docID, 0, 0, []byte(`{"datastore":"fromDs"}`), 0)
+				require.NoError(t, err)
+				return cas
+			},
+		},
+		{
+			name:  "dest doc newer, expect keep old version",
```
Comment (adamcfraser) [inline xdcr/xdcr_test.go]: This says dest doc newer, but looks like fromDs gets written to last?

File: xdcr/xdcr_test.go [hunk 2]
```diff
@@ -50,74 +52,299 @@ func TestMobileXDCRNoSyncDataCopied(t *testing.T) {
...
-					var vvXattrVal map[string]any
-					require.NoError(t, base.JSONUnmarshal(vvXattrBytes, &vvXattrVal))
-					assert.NotNil(c, vvXattrVal[version])
-					assert.NotNil(c, vvXattrVal[source])
-					assert.NotNil(c, vvXattrVal[curCAS])
-
-				}, time.Second*5, time.Millisecond*100)
-			}
-		}
+func TestReplicateVV(t *testing.T) {
```
Comment (adamcfraser) [inline xdcr/xdcr_test.go:149]: This test looks like what's it's actually testing is XDCR's LWW logic?  Should it be renamed TestReplicateLWW?

File: xdcr/xdcr_test.go [hunk 3]
```diff
@@ -50,74 +52,299 @@ func TestMobileXDCRNoSyncDataCopied(t *testing.T) {
...
+	require.JSONEq(t, ver1Body, string(body))
+	require.Contains(t, xattrs, base.VvXattrName)
+	casString := string(base.Uint64CASToLittleEndianHex(fromCAS))
+	var vv *db.HybridLogicalVector
+	require.NoError(t, base.JSONUnmarshal(xattrs[base.VvXattrName], &vv))
+	require.Equal(t, &db.HybridLogicalVector{
+		CurrentVersionCAS: casString,
+		SourceID:          fromBucketSourceID,
+		Version:           casString,
+	}, vv)
```
Comment (adamcfraser) [inline xdcr/xdcr_test.go]: This seems like commonly reused code.  Maybe add a toDs.VerifyCV(t, cvcas, sourceID, version) function?  (which could call a toDS.VerifyHLV(t, HLV) for future tests that care about pv?

File: xdcr/rosmar_xdcr.go [hunk 4]
```diff
@@ -186,30 +193,55 @@ func (r *rosmarManager) Stop(_ context.Context) error {
 	return nil
 }
 
-// opWithMeta writes a document to the target datastore given a type of Deletion or Mutation event with a specific cas.
-func opWithMeta(ctx context.Context, collection *rosmar.Collection, originalCas uint64, event sgbucket.FeedEvent) error {
-	var xattrs []byte
+// opWithMeta writes a document to the target datastore given a type of Deletion or Mutation event with a specific cas. The originalXattrs will contain only the _vv and _mou xattr.
```
Comment (adamcfraser) [inline xdcr/rosmar_xdcr.go:196]: XDCR will be preserving the _sync xattr on the target, so I had anticipated originalXattrs would include that for setWithMeta to work as expected?  Or is that pending work for a later ticket.

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: xdcr/rosmar_xdcr.go (hunk 4 above)
Comment (torcolvin) [reply xdcr/rosmar_xdcr.go:196]: https://jira.issues.couchbase.com/browse/CBG-4281 for doing this, even though I think it's pretty trivial.

File: xdcr/xdcr_test.go (hunk 1 above)
Comment (torcolvin) [reply xdcr/xdcr_test.go]: This test was wrong, thanks.

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**